Twine 1 TWS documentation

videlais opened the pull request:
Apologies for the multiple PRs on this. I had to re-name this branch and fix an issue with translating out of the Python date stamp.

---

Original:

When preparing this, I noticed you mentioned in the Twine-Utils codebase how some of the properties were not documented (see [Story.ts](https://github.com/klembot/twine-utils/blob/main/src/story.ts#L30)). Is there any past documentation we could point to like the revived [Twee 1/Twine 1](http://twinery.org/1/doc/#twine,story) docs, [@klembot](https://github.com/klembot)?

Updated:

~~The only two properties I am unsure of are created and modified per passage. In testing, these register as "PFunction" (when using Twine-Utils). Testing in another library produced a date stamp array with nine entries, but obviously the "Function" type does not translate into JSON.~~

Python date stamps are strange. While using [pickleparser](https://github.com/ewfian/pickleparser) (what Twine-Utils uses), both `modified` and `created` will be correctly detected as 'PFunction' objects. However, this means they will not be rendered into JSON. Accessing the live `args` of each function shows the internal Python date stamp as an array. I have tried to better inform people attempting to read this data in the specification, but I'm also very open to a better way to explain and help people parse things.

videlais commented:
> I genuinely don't remember much about the file format.

Heh. Very fair. I am asking about something that started, what? Over a decade ago. Won't Twee/Twine hit 15 years old next summer? Wild.

As far as I know, the Twine archive on IFDB only dates back to 2013. Looking through the files shows  _[Hallowmoor](https://ifdb.org/viewgame?id=65ifglmwo3f2otj)_ from November 2013, which contains a TWS file. ([Hallowmoor_source.tws.zip](https://github.com/iftechfoundation/twine-specs/files/12515519/Hallowmoor_source.tws.zip)). Using the online version of [Pickle Parser](https://ewfian.github.io/pickleparser/) shows the story `metadata` and `storyPanel.snapping` properties do not exist, which means they were probably added after this file came out. (Twine 1.4.0 was published in [December 2013](https://web.archive.org/web/20131216041024/http://twinery.org/wiki/twine_1.4_release_notes)).

Unless we have a TWS to test before 2013, I think this is pretty good? The only thing I am not as sure on is the parsing of the Python date stamp data. I've been considering setting up Python to see if I can deserialize the data to confirm everything, but I have not taken that step yet.

videlais commented:
After installing and preparing some test code in Python today, I was finally able to verify I was wrong across multiple keys of the Python time data in the last commit. Twine 1 used [`time.struct_time`](https://docs.python.org/3/library/time.html#time.struct_time), which means the last attribute (or entry in an array in JavaScript) represented the daylight saving time offset.

I've added a new subsection explaining this formatting to the file in a new commit.

tmedwards reviewed (commented):
My only comment here would be that Twine 1 uses Python v2, rather than v3.  Your links may want to reflect that—you have at least one link to v3 documentation in there.

videlais commented:
> My only comment here would be that Twine 1 uses Python v2, rather than v3. Your links may want to reflect that—you have at least one link to v3 documentation in there.

That's a great point. I've updated two links in the latest commit to use Python 2 documentation in reference to Twine 1 and added text to explain the older version.

videlais commented:
@tmedwards and @klembot: Is there any remaining feedback or changes for this documentation? Is there anything else we should capture or explain?

videlais commented:
@tmedwards: Any additional changes before I merge this?

I agree with Klembot that we can always add or adjust things in the future, should that ever be needed.

tmedwards commented:
Sorry.  Missed the ping.
